benchexec/container.py

```diff
@@ -723,28 +763,130 @@ def make_overlay_mount(mount, lower, upper, work):
         work,
     )
 
-    def escape(s):
-        """
-        Safely encode a string for being used as a path for overlayfs.
-        In addition to escaping ",", which separates mount options,
-        we need to escape ":", which overlayfs uses to separate multiple lower dirs
-        (cf. https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt).
-        """
-        return s.replace(b"\\", rb"\\").replace(b":", rb"\:").replace(b",", rb"\,")
-
     libc.mount(
         b"none",
         mount,
         b"overlay",
         0,
         b"lowerdir="
-        + escape(lower)
+        + escape_overlayfs_parameters(lower)
+        + b",upperdir="
+        + escape_overlayfs_parameters(upper)
+        + b",workdir="
+        + escape_overlayfs_parameters(work),
+    )
+
+
+def check_use_fuse_overlayfs(mount_base, dir_modes):
+    """
+    Check whether an overlay mountpoint requires the use of fuse-overlayfs
+    by determining if there are any sub-mounts below it.
+    """
+    mount_points = [
+        (full_mountpoint, fstype)
+        for _unused_source, full_mountpoint, fstype, _options in get_mount_points()
+        if util.path_is_below(full_mountpoint, mount_base)
+    ]
+
+    for full_mountpoint, fstype in mount_points:
+        mountpoint = full_mountpoint[len(mount_base) :] or b"/"
+        mode = determine_directory_mode(dir_modes, mountpoint, fstype)
+
+        if not mode or not os.path.exists(mountpoint):
+            continue
+
+        if mode == DIR_OVERLAY:
+            # Check if there are any sub-mounts within the current overlay mount point
+            for sub_mountpoint, _unused_fstype in mount_points:
+                if (
+                    util.path_is_below(sub_mountpoint, mountpoint)
+                    and sub_mountpoint != mountpoint
+                ):
+                    return True
+
+    return False
+
+
+@contextlib.contextmanager
+def permitted_cap_as_ambient():
```

Comment:
Please move the documentation here and explain from a caller side what this method does.